### ISSUE TITLE
GOLDEN DICTIONARY IN UBUNTU 20.04 UNEXPECTED RAM USAGE OR MEMORY LEAK

### ISSUE THREAD

**Issue by therickyrick** [goldendict/goldendict#1332]
I am using this to look up words meaning by selecting the words from the pdf document, initially the RAM usage is 30-40 MB, but thereafter for every 1 word being searched RAM usage goes to 6-10 MB per word, after 30 min the RAM usage is around 1 GB by Gloden dict.

I am using the scan popup mode which displays the meaning of the word as soon as the word in highlighted in the pdf document with around 8-10 dictionaried added. I assume this as a memory leak or some programming issue.

Rest everything is ok, if this problem gets resolved i will definitely give it 5 stars. 

You can reach me on divya123sweety1@gmail.com for more light on this issue.

Thanks in advance.

**Comment by vedgy** [comment]
I can reproduce the RAM usage growth on my Manjaro Xfce system. But it's 3-4 MB per word on average here. Also RAM usage does not always grow. It grows or shrinks with no obvious pattern after each translated word. On average though, it grows over time.

**Comment by nonwill** [comment]
Replace sptr with std::shared_ptr from C++11 will fix some leaks and avoid crashing the app in some case，just refer to https://github.com/goldendict/goldendict/pull/1129/commits/8b93c25e7af53ae94c4ffeadd18402796136fe8f. 
But the program really have more codes leading to memory leaks --- Totally review all the codes，memory leaks probably occur with some type of dictionary processing，and the exception handling also issues some...... And also, if it's based on Qt 5.12(and 5.15), QObject orphaned connections may make memory  [soft](https://bugreports.qt.io/browse/QTBUG-88248)-[leaks](https://bugreports.qt.io/browse/QTBUG-87774)......

**Comment by vedgy** [comment]
There are automated tools that detect memory leaks. For example, [Valgrind](https://valgrind.org/docs/manual/quick-start.html#quick-start.intro), [GCC's `-fsanitize=leak](https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html), [Clang LeakSanitizer](https://clang.llvm.org/docs/LeakSanitizer.html). I guess no one bothered to test GoldenDict with them for a long time, if ever.

**Comment by vedgy** [comment]
> Replace sptr with std::shared_ptr from C++11 will fix some leaks and avoid crashing the app in some case，just refer to 8b93c25.

Why? Because of a separation of `new` and `sptr` constructor (lack of `std::make_shared()`), because of thread un-safety (non-atomic reference counter) or is there a bug in `sptr`'s implementation?

> and the exception handling also issues some

I can understand that. Mixing Qt and liberal throwing of exceptions is uncommon and thus risky.

**Comment by nonwill** [comment]
> because of thread un-safety (non-atomic reference counter)

Yes, it's the risk, for the memory may be never freed(deleted) which makes a leak, or the raw pointer may be double-deleted which crashes the app.

>std::shared_ptr is not a perfect

Yes. I _usually_ prefer the raw pointer. But here std::shared_ptr would be more _compatible_ than std::unique_ptr to replace sptr, and it needs less codes too.
